### DISCUSSION TITLE
Forcing Custom Control fill available space in the container if either width or height is not set

### DISCUSSION THREAD

**Discussion by fitdev** [AvaloniaUI/Avalonia#13187] [Q&A] [answered]
I have a custom (owner-draw) control that inherits from `Control`. I want to enable the behavior such that when it is placed inside of a container like a Panel, it should _fill all available space_, i.e. such that its Width and Height would be set to match that of available space (provided of course they are not explicitly specified).

I have come up with the following way to do this:

```cs
  protected override Size MeasureOverride(Size availableSize) {
    if (double.IsNaN(Height)) Height = availableSize.Height;
    if (double.IsNaN(Width)) Width = availableSize.Width;
    return availableSize;
  }
```

But I am not sure if this is correct and/or future-proof, as it relies on the current implementation of `Layoutable.MeasureCore` in Avalonia. I mean setting Width and Height in a MeasureOverride does not seem like a good idea. But on the other hand, how else can I accomplish what I want?

**Comment by maxkatz6** [comment, marked as answer]
> it is placed inside of a container like a Panel, it should fill all available space

Normally, it's enough to just set HorizontalAlignment to Stretch and the same for vertical one.

But IIRC alignment properties don't affect Measure step, as in measure step we want to know size of the control itself, which can be bigger or smaller than parent available size. 
And in Arrange step we respect alignment properties, adjusting control size to property fit in the parent control.

**Reply by maxkatz6** [reply]
> it is placed inside of a container like a Panel, it should fill all available space

Also, it's mostly up to the container. For example, if you have StackPanel (Orientation=Vertical), vertical available size won't be constrained, and control cannot be stretched vertically.

**Reply by fitdev** [reply]
Thank you for your explanation. I certainly tried setting `HorizontalAlignment` and `VerticalAlignment` to `Stretch` in XAML in various containers - from `DockPanel` (with `LastChild` = `Fill`), to regular `Panel`, to `Border`, and it did not work anywhere (interestingly the `Bounds` were reported properly as filling the whole container, yet `Width` and `Height` were not set and so the control was "invisible") until I added the `MeasureOverride` above.

So then I take it that my current approach with `MeasureOverride` is ok then? I mean I was not sure if it's appropriate to actually set `Width` and `Height` in this routine.

**Reply by fitdev** [reply]
Actually, I figured it out: I should have been using `Bounds.Width` and `Bounds.Height` in my control's rendering logic instead of `Width` and `Height`. If I use `Bounds` then there is no need for `MeasureOverride` at all.
